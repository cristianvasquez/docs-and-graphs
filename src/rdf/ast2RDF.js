import rdf from 'rdf-ext'
import ns from '../namespaces.js'

function ast2RDF ({ simpleAst }) {

  const dataset = rdf.dataset()

  const root = rdf.blankNode()
  const pointer = rdf.clownface({ term: root, dataset })

  triplifyNode({ pointer, nodeUri: root, json: simpleAst })
  return pointer

}

function triplifyNode ({ pointer, nodeUri, json }) {
  if (json.type) {
    pointer.node(nodeUri).addOut(ns.rdf.type, ns.mark[json.type])

  }
  for (const child of json.children ?? []) {
    const childUri = rdf.blankNode()
    triplifyNode({ pointer, nodeUri: childUri, json: child })
  }
  return pointer
}

// function createItem ({ pointer, text, type }) {
//   const newUri = rdf.blankNode()
//   pointer.node(newUri).addOut(ns.rdf.type, type)
//   pointer.node(newUri).addOut(ns.mark.value, rdf.literal(text))
//   augment(pointer.node(newUri), text)
//   return newUri
// }

// function augment (pointer, text) {
//   for (const tag of extractTags(text)) {
//     pointer.addOut(ns.mark.tag, rdf.literal(tag))
//   }
//   for (const { property, value } of extractInlineFields(text)) {
//     const propValue = rdf.blankNode()
//     pointer.addOut(ns.mark.inline, propValue)
//     pointer.node(propValue).addOut(ns.rdf.type, ns.mark.Inline)
//     pointer.node(propValue).
//       addOut(ns.mark.inlineProperty, rdf.literal(property))
//     pointer.node(propValue).addOut(ns.mark.inlineValue, rdf.literal(value))
//
//   }
// }

// function createList (pointer, strings) {
//   const listUri = rdf.blankNode()
//   pointer.node(listUri).addOut(ns.rdf.type, ns.mark.List)
//   for (const text of strings) {
//     const itemUri = createItem({ pointer, text, type: ns.mark.ListItem })
//     push(pointer.node(listUri), ns.mark.nodes, itemUri)
//   }
//   return listUri
// }

export { ast2RDF }
