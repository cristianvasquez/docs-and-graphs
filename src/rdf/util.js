function pushToList (pointer, predicate, term) {
  if (pointer.out(predicate).isList()) {
    const elements = [...pointer.out(predicate).list(), term]
    pointer.deleteList(predicate)
    pointer.addList(predicate, elements)
  } else {
    pointer.addList(predicate, [term])
  }
}

export { pushToList }
