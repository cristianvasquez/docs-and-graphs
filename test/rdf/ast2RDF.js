import { expect } from 'expect'

import { toJson } from '../../index.js'

import toMatchSnapshot from 'expect-mocha-snapshot'
import { ast2RDF } from '../../src/rdf/ast2RDF.js'
import { prettyPrint } from '../util.js'
import tests from '../tests.js'

expect.extend({ toMatchSnapshot })

describe('astDag', async function () {
  for (const current of tests) {
    it(current.title, async function () {
      const simpleAst = toJson(current.markdown)
      const pointer = ast2RDF({ simpleAst })

      const turtle = await prettyPrint(pointer.dataset)
      expect(turtle).toMatchSnapshot(this)
    })
  }
})
