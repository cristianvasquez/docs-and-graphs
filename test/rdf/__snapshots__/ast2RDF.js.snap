// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`astDag Headings 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

_:b3 a mark:root .

"
`;

exports[`astDag Links 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

_:b1 a mark:root .

"
`;

exports[`astDag Outline 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

_:b2 a mark:root .

"
`;

exports[`astDag People 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

_:b4 a mark:root .

"
`;
